### ISSUE TITLE
OpenCV OpenCL support?

### ISSUE THREAD

**Issue by tensorbuffer** [bytedeco/javacpp-presets#1162]
Hello,
I first did opencv java build, with OpenCL support, on Qualcomm device. I found out that does not work, because opencv java build does not expose UMat, so in the java code I can't use UMat. Reference: https://github.com/opencv/opencv/issues/9154
From @saudet 's comment I came to know this javacv / javacpp-presets project. However after I downloaded the opencv-4.0.1-1.4.4.jar file and extract, I also do not see UMat (under org/opencv/core folder, where you can see Mat.class)
I wonder how JavaCV (or javacpp) can enable OpenCL support on device?
If my understanding is correct, UMat is a must for opencv, for example: https://github.com/opencv/opencv/blob/4.x/modules/imgproc/src/morph.dispatch.cpp#L1174 Each CV_OCL_RUN() has a check for _dst.isUMat().
Pls let me know, if this approach does not work we need to move the algorithm to native layer.

**Comment by saudet** [comment]
UMat is available: http://bytedeco.org/javacpp-presets/opencv/apidocs/org/bytedeco/opencv/opencv_core/UMat.html

Which functions are you looking to call?

**Comment by tensorbuffer** [comment]
Thanks, there are many, one example is morphologyEx().
I wonder why the one I downloaded does not contain? https://mvnrepository.com/artifact/org.bytedeco.javacpp-presets/opencv/4.0.1-1.4.4 this is the latest version already.

**Comment by saudet** [comment]
That function is available yes: http://bytedeco.org/javacpp-presets/opencv/apidocs/org/bytedeco/opencv/global/opencv_imgproc.html#morphologyEx-org.bytedeco.opencv.opencv_core.UMat-org.bytedeco.opencv.opencv_core.UMat-int-org.bytedeco.opencv.opencv_core.UMat-org.bytedeco.opencv.opencv_core.Point-int-int-org.bytedeco.opencv.opencv_core.Scalar-

I'm pretty sure it's also available in OpenCV 4.0.1, but why do you need to use such an old version of OpenCV?

**Comment by tensorbuffer** [comment]
Oh I am fine with any latest opencv version, I only found the 4.0.1 downloadable binaries. Where I can find the latest ones (opencv 4.5.5?) that contains the 1.5.7 javacpp-presets if any?

**Comment by saudet** [comment]
Check the download page: http://bytedeco.org/download/

**Comment by tensorbuffer** [comment]
Thanks! I downloaded through the GitHub releases, unzipped and got javacpp-presets-platform-1.5.7-bin/opencv-android-arm64.jar. Now how to use it together with JavaCV?

**Comment by saudet** [comment]
If you're using Android Studio, you can start from this sample project:
https://github.com/bytedeco/sample-projects/tree/master/JavaCV-android-example

**Comment by tensorbuffer** [comment]
Well, I didn't see any of the jar file contains org.bytedeco.opencv, only org.bytedeco.javacv

**Comment by saudet** [comment]
Well, I'm not going to teach you all about Maven and Gradle, but this is a good place to start I think:
https://docs.gradle.org/current/userguide/dependency_management.html

**Comment by saudet** [comment]
That's what Maven and Gradle do, read up.

**Comment by tensorbuffer** [comment]
Do you mean that I shouldn't use the releases from GitHub, instead I should use Marvin? I am not sure if that's allowed in our build system (accessing external repository), big company has strict rules... I was hoping to do a quick prototype to see if this approach works or not.

**Comment by saudet** [comment]
You can put all the JAR files in your class path, sure, that works too.

**Comment by tensorbuffer** [comment]
Ok I figured out that I didn't copy over the opencv.jar.
Now my problem is the add() subtract() multiply functions.
I didn't see them in http://bytedeco.org/javacpp-presets/opencv/apidocs/overview-summary.html
Are they supported?

**Comment by saudet** [comment]
Yeah, there's lots of those: http://bytedeco.org/javacpp-presets/opencv/apidocs/index-all.html

**Comment by tensorbuffer** [comment]
Since the current app is written in Kotlin, I found there's issue that hard to resolve, for example there's code to convert a DoubleArray to Scalar, something like Scalar(DoubleArray(...)). And I got this error:
None of the following functions can be called with the arguments supplied: 
public constructor Scalar(p0: Double) defined in org.bytedeco.opencv.opencv_core.Scalar
public constructor Scalar(p0: Pointer!) defined in org.bytedeco.opencv.opencv_core.Scalar
public constructor Scalar(p0: Scalar!) defined in org.bytedeco.opencv.opencv_core.Scalar

Looks like DoubleArray is Kotlin specific, so it's hard to use JavaCV with it... There are many many other compile errors and I guess it will take quite some time to change the code (not sure if feasible though)

**Comment by saudet** [comment]
Kotlin is fully compatible with Java. You'll definitely not have any issues because of Kotlin.

**Comment by tensorbuffer** [comment]
I think the problem is:
opencv Scalar has "public Scalar​(double[] vals)"
https://docs.opencv.org/3.4/javadoc/org/opencv/core/Scalar.html

javacpp scalar does not have this (array of double), it only has double /Pointer / Scalar
http://bytedeco.org/javacpp-presets/opencv/apidocs/org/bytedeco/opencv/opencv_core/Scalar.html

**Comment by tensorbuffer** [comment]
Another question: for Mat, I should use org.bytedeco.opencv.opencv_core, not org.opencv.core, right? the opencv one has submat() function but not in bytedeco.

**Comment by saudet** [comment]
The function for that in C++ API is `operator()()`, but it doesn't give it any name. JavaCPP maps it to `apply()`:
http://bytedeco.org/javacpp-presets/opencv/apidocs/org/bytedeco/opencv/opencv_core/Mat.html#apply-org.bytedeco.opencv.opencv_core.Range-org.bytedeco.opencv.opencv_core.Range-

**Comment by tensorbuffer** [comment]
ok thanks. And also missing MatofPoint().

**Comment by saudet** [comment]
The C++ API doesn't have that, just use `Mat`, we can use it for points as well.

**Comment by tensorbuffer** [comment]
Actually, how hard is it to add UMat support in opencv java build? Where to start and which code in javacv can be a reference point? I am thinking maybe that's a quicker prototype solution...

**Comment by saudet** [comment]
@alalek answered your question on issue  https://github.com/opencv/opencv/issues/9154#issuecomment-354566824 you linked to. Anyway, the Java API wasn't designed for performance, so if that's important for your application you shouldn't be using it in any case.

**Comment by tensorbuffer** [comment]
"the Java API wasn't designed for performance" --- yes performance is important to us. Why performance would be an issue, is it due to the JNI layer cost? I thought it's just a bunch of java layer API call to native layer?

**Comment by saudet** [comment]
Well, there's a lot of issues with the JNI code from the Java API, which is bundled as is in JavaCV, so it's the same thing, yes. There's "good" code and then there's "bad" code, it's not specific to Java or anything. Java just isn't a priority for the team working on OpenCV.
